src/com/ms/silverking/cloud/dht/daemon/storage/convergence/ChecksumTreeServer.java

```diff
@@ -131,177 +130,180 @@ public ChecksumNode getRegionChecksumTree_Local(ConvergencePoint cp, RingRegion
-            if (verbose || debug) {
-                System.out.printf("null root for %x\t%s\t%s\n", ns, mixedCP, region);
-            }
-        }
-        if (debug) {
-            System.out.println("computed:");
-            System.out.println(root);
-        }
-        bufferSize = estimateBufferSize(root);
+  }
+
+  private ChecksumTreeGroup computeChecksumTreeGroup(RingRegion region, long minVersion, long maxVersion) {
+    return computeChecksumTreeGroup(ImmutableSet.of(region), minVersion, maxVersion, true);
+  }
+
+  /**
+   * Computes a new ChecksumTreeGroup given a version. Uses the regions in the ringMaster.
+   *
+   * @param ringID     TODO
+   * @param minVersion
+   * @param maxVersion
+   * @return
+   */
+  private ChecksumTreeGroup computeChecksumTreeGroup(Collection<RingRegion> regions, long minVersion, long maxVersion,
+      boolean allowRegionNotFound) {
+    Log.info("computeChecksumTreeGroup w/ regions");
+    if (debug) {
+      System.out.printf("computeChecksumTreeGroup w/ regions %x\n", ns);
+    }
+    if (regions != null) {
+      nsStore.readLockAll();
+      try {
+        return TreeBuilder.build(regions, nsStore.keyAndVersionChecksumIterator(minVersion, maxVersion), entriesPerNode,
+            nsStore.getTotalKeys(), absMillisTimeSource.absTimeMillis(), minVersion, maxVersion, allowRegionNotFound);
+      } finally {
+        nsStore.readUnlockAll();
+      }
+    } else {
+      return null;
+    }
+  }
+
+  //////////////////////////////////////////////////////////////////////
+  // Handle incoming request for a convergence tree
+
+  public void getChecksumTree(UUIDBase uuid, ConvergencePoint targetCP, ConvergencePoint sourceCP,
+      MessageGroupConnection connection, byte[] originator, RingRegion region) {
+    ChecksumNode root;
+    ProtoChecksumTreeMessageGroup pmg;
+    int bufferSize;
+    ConvergencePoint mixedCP;
+
+    if (verbose || debug) {
+      System.out.printf("getChecksumTree %x\t%s\t%s\t%s\t%s\n", ns, targetCP, sourceCP, region, uuid);
+    }
+    // FUTURE - support both min and max version
+
+    // We create a mixedCP because we need to retrieve target data from the source regions, but
+    // we need to use the target data version
+    mixedCP = sourceCP.dataVersion(targetCP.getDataVersion());
+    root = getRegionChecksumTree_Local(mixedCP, region, new LongInterval(Long.MIN_VALUE, mixedCP.getDataVersion()));
+    if (root == null) {
+      if (verbose || debug) {
+        System.out.printf("null root for %x\t%s\t%s\n", ns, mixedCP, region);
+      }
+    }
+    if (debug) {
+      System.out.println("computed:");
+      System.out.println(root);
+    }
+    bufferSize = estimateBufferSize(root);
+    pmg = null;
+    do {
+      try {
+        pmg = new ProtoChecksumTreeMessageGroup(uuid, ns, targetCP, originator, root, bufferSize);
+        //pmg = new ProtoChecksumTreeMessageGroup(uuid, ns, mixedCP, originator, root, bufferSize);
+      } catch (BufferOverflowException bfe) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)